### PR TITLE
flight & python: file transfer functionality

### PR DESCRIPTION
**Pull request by mlyle**
And assorted small cleanups.

If/when architecture for this is approved, will set about A) GCS implementation, and B) updating uavtalk protocol docs.

### CODE REVIEW

**Comment by mlyle** [comment]
Tested on Sparky1 & Sparky2.  Gets about 10k/sec in transfer over USB HID, which is not blazingly fast.

OTOH we put 4 in bInterval, which means 250 messages/sec.  At 64 bytes payload each, maximum theoretical transfer rate is 63 (len prefixed) * 250 = 15750 bytes/sec, before any of the uavtalk overhead, channel turnaround time, etc.

File: python/dronin/uavtalk.py [hunk 1]
```diff
@@ -171,7 +171,9 @@ def process_stream(uavo_defs, use_walltime=False, gcs_timestamps=None,
         # Search for object.
         uavo_key = '{0:08x}'.format(objId)
         if not uavo_key in uavo_defs:
-            #print "Unknown object 0x" + uavo_key + "type = ", pack_type
+            #print "Unknown object 0x" + uavo_key + " type = ", pack_type
+            obj_len = pack_len - MIN_HEADER_LENGTH
```
Comment (tracernz) [inline python/dronin/uavtalk.py]: This is defeated at line 192 (and it looks more correct)?

File: flight/Libraries/inc/uavtalk.h [hunk 2]
```diff
@@ -31,8 +31,10 @@
 #define UAVTALK_H
 
 // Public types
-typedef int32_t (*UAVTalkOutputCb)(void *ctx, uint8_t* data, int32_t length);
+typedef int32_t (*UAVTalkOutputCb)(void *ctx, uint8_t *data, int32_t length);
```
Comment (tracernz) [inline flight/Libraries/inc/uavtalk.h:34]: :+1:

File: flight/Libraries/inc/uavtalk_priv.h [hunk 3]
```diff
@@ -88,24 +98,28 @@ typedef struct {
...
+	UAVTalkFileCb fileCb;
 	void *cbCtx;
 } UAVTalkConnectionData;
 
 #define UAVTALK_CANARI         0xCA
 #define UAVTALK_WAITFOREVER     -1
 #define UAVTALK_NOWAIT          0
 #define UAVTALK_SYNC_VAL       0x3C
-#define UAVTALK_TYPE_MASK      0x78
+#define UAVTALK_TYPE_MASK      0x70
```
Comment (tracernz) [inline flight/Libraries/inc/uavtalk_priv.h:109]: Seems this is really UAVTALK_VER_MASK. :/

File: flight/Libraries/uavtalk.c [hunk 4]
```diff
@@ -254,6 +261,26 @@ UAVTalkRxState UAVTalkProcessInputStreamQuiet(UAVTalkConnection connectionHandle
 		if (iproc->rxCount < 4)
 			break;
 
+		if (iproc->type == UAVTALK_TYPE_FILEREQ) {
```
Comment (tracernz) [inline flight/Libraries/uavtalk.c:264]: Do we want support for receiving files on the flight side? Could be useful for PicoC replacement if it ever eventuates. Might be safer to handle it here and error out in any case.

File: flight/Libraries/uavtalk.c [hunk 5]
```diff
@@ -617,6 +640,95 @@ static int32_t sendBuf(UAVTalkConnection connectionHandle, uint8_t *buf, uint16_
...
+			total_len += cb_numbytes;
+
+			file_offset += cb_numbytes;
+
+			if (i == 5) {	/* 6 messages per chunk */
+				resp->flags = UAVTALK_FILEDATA_LAST;
+			} else {
+				resp->flags = 0;
+			}
+		} else {
```
Comment (tracernz) [inline flight/Libraries/uavtalk.c:701]: So if no callback registered or an error getting the file, 0 bytes response with EOF.. seems reasonable. ENOENT could also be nice to have?

File: flight/Modules/Telemetry/telemetry.c [hunk 6]
```diff
@@ -393,6 +416,67 @@ static void addAckPending(telem_t telem, UAVObjHandle obj, uint16_t inst_id)
...
+ * id to the actual unit of information, and returns/copies it.
+ * Currently only operates on partitions.
+ *
+ * \param[in] ctx Callback context (telemetry subsystem handle)
+ * \param[in] file_id The requested file_id
+ * \param[in] offset The offset of information requested
+ * \param[in] len The maximum amount of information to return.  A short read
+ * is permissable, as long as some bytes are returned.
+ * \returns The number of bytes filled, 0 on EOF, negative on error.
+ */
```
Comment (tracernz) [inline flight/Modules/Telemetry/telemetry.c:430]: Nice work on all the Doxygen. :grin:

File: python/dronin/telemetry.py [hunk 7]
```diff
@@ -193,23 +197,70 @@ def __handle_handshake(self, obj):
 
             if obj.Status == obj.ENUM_Status['Disconnected']:
                 # Request handshake
-                print("Disconnected")
+                #print("Disconnected")
```
Comment (tracernz) [inline python/dronin/telemetry.py:200]: These should probably all use python logging so they can be enabled at runtime... For another PR. :)

File: python/dronin/uavtalk.py [hunk 8]
```diff
@@ -170,9 +173,12 @@ def process_stream(uavo_defs, use_walltime=False, gcs_timestamps=None,
         # Search for object.
         uavo_key = '{0:08x}'.format(objId)
         if not uavo_key in uavo_defs:
-            #print "Unknown object 0x" + uavo_key
+            #print "Unknown object 0x" + uavo_key + " type = ", pack_type
+            obj_len = pack_len - MIN_HEADER_LENGTH
```
Comment (tracernz) [inline python/dronin/uavtalk.py]: Isn't this overritten at line 195? header_fmt.size seems better too.

File: flight/Libraries/inc/uavtalk_priv.h (hunk 3 above)
Comment (mlyle) [reply flight/Libraries/inc/uavtalk_priv.h:109]: Heh yup.  And no idea why it was malformed this way.

File: flight/Libraries/uavtalk.c (hunk 4 above)
Comment (mlyle) [reply flight/Libraries/uavtalk.c:264]: Yes, the plan is to allow pushing files the other way, though the state machine of the initiator is slightly trickier.

File: flight/Libraries/uavtalk.c (hunk 5 above)
Comment (mlyle) [reply flight/Libraries/uavtalk.c:701]: Yes flags exist to extend this.

File: flight/Modules/Telemetry/telemetry.c (hunk 6 above)
Comment (mlyle) [reply flight/Modules/Telemetry/telemetry.c:430]: Thx :)

File: python/dronin/telemetry.py (hunk 7 above)
Comment (mlyle) [reply python/dronin/telemetry.py:200]: Yah, these were problematic-- corrupted my received files.  Ditto for the bad CRC messages.

**Pull request merged**